### ISSUE TITLE
Incorrect padding on simple tables

### ISSUE THREAD

**Issue by markhildreth** [RyanBluth/term-table-rs#4]
I noticed that the default padding of columns for simple tables are not what I expected. In many of the examples. Here is a test case that shows this:

```rust
#[test]
fn correct_default_padding() {
    let mut table = Table::new();
    table.separate_rows = false;
    table.style = TableStyle::simple();
    table.add_row(Row::new(vec![
        TableCell::new_with_alignment("A", 1, Alignment::Center),
        TableCell::new_with_alignment("B", 1, Alignment::Center),
    ]));
    table.add_row(Row::new(vec![TableCell::new(1), TableCell::new("1")]));
    table.add_row(Row::new(vec![TableCell::new(2), TableCell::new("10")]));
    table.add_row(Row::new(vec![TableCell::new(3), TableCell::new("100")]));
    let expected = r"+---+-----+
| A |  B  |
| 1 | 1   |
| 2 | 10  |
| 3 | 100 |
+---+-----+
";
    println!("{}", table.render());
    assert_eq!(expected, table.render());
}
```

Here's what the output actually is:

```
+-----+-------+
|  A  |   B   |
| 1   | 1     |
| 2   | 10    |
| 3   | 100   |
+-----+-------+
```

To me, there is more padding than is necessary. The first column has a width of five characters, where three would do. The second column has a width of seven characters, where five would do. To me, the expected output should be:

```
+---+-----+
| A |  B  |
| 1 | 1   |
| 2 | 10  |
| 3 | 100 |
+---+-----+
```

I took a whack at fixing this myself; it appears that the problem is that width is taken into account twice (see my attempt at a fix [here](https://github.com/RyanBluth/term-table-rs/compare/master...markhildreth:fix-padding-calculated-twice?expand=1)). The fix I showed works, but then it causes a change of output and regression with more complicated tables. I decided at that point to stop and just report the issue.

**Comment by RyanBluth** [comment]
Thanks for reporting the issue. I'll look into this

**Comment by RyanBluth** [comment]
I think I have this all sorted out. I've pushed the changes to master if you want to give it a try

**Comment by markhildreth** [comment]
Looks better. I also just noticed that my sample test included an extra seperator under the header, although it wasn't strictly supposed to be there. Anyway, assuming I removed that seperator, master seems to have fixed it.

I did notice that this example is still off:

```
    let mut table = Table::new();
    table.separate_rows = false;
    table.style = TableStyle::simple();
    table.add_row(Row::new(vec![
        TableCell::new_with_alignment("A1", 1, Alignment::Center),
        TableCell::new_with_alignment("B", 1, Alignment::Center),
    ]));
    println!("{}", table.render());
```
Expected:
```
+----+---+
| A1 | B |
+----+---+
```
Actual (v1.2.1, correct centering but extra padding):
```
+------+-----+
|  A1  |  B  |
+------+-----+
```
Actual (master, incorrect padding):
```
+-----+---+
|  A1 | B |
+-----+---+
```

**Comment by RyanBluth** [comment]
Good catch. I've fixed that issue as well now. 
I've realized there is a bit of a chicken and egg problem though. Depending on the content of the cells, it's essentially impossible to have everything centered. If a cell spans two columns and requires an even cell width to be centered properly then a subsequent row of two cells cannot have the divider in the middle if their content is the same length. Not sure what the optimal solution is

**Comment by markhildreth** [comment]
What you have in master looks good to me for the simple tables I've tested on.

You are correct; at some point there isn't an optimal solution to centering, since even and odd column lengths cannot be centered with the same amount of padding. For example, there is no "correct" way to make everything in the column containing the values `["ODD", "EVEN"]` centered.

```
+-------+
|  ODD  |
|  EVEN |
| EVEN  | // Alternatively to above
+-------+

+------+
| ODD  |
|  ODD | // Alternatively to above
| EVEN |
+------+
```

I'll leave it up to you as to how far down the rabbit hole you want to go :) This isn't blocking me on anything, I was checking out the library and thought I'd leave the issue. Here's a nested one that theoretically has an "optimal" solution, in that there is a way to ensure that any cell which wants to be centered can be centered with the minimum amount of padding needed.

```
fn main() {
    let mut table = Table::new();
    table.separate_rows = false;
    table.style = TableStyle::simple();
    table.add_row(Row::new(vec![
        TableCell::new_with_alignment("A", 1, Alignment::Center),
        TableCell::new_with_alignment("B", 1, Alignment::Center),
    ]));
    table.add_row(Row::new(vec![TableCell::new(1), TableCell::new("1")]));
    table.add_row(Row::new(vec![TableCell::new(2), TableCell::new("10")]));
    table.add_row(Row::new(vec![TableCell::new(3), TableCell::new("100")]));
    table.add_row(Row::new(vec![TableCell::new_with_alignment("Spanner", 2, Alignment::Center)]));
    let expected = "
+---+-----+
| A |  B  |
| 1 | 1   |
| 2 | 10  |
| 3 | 100 |
| Spanner |
+---------+
";
    println!("{}", table.render());
    assert_eq!(expected.trim(), table.render().trim());
}
```

**Comment by RyanBluth** [comment]
I've published 1.3.0 with the improvements I've made so far. I'll close this issue for now since the original issue you raised has been resolved, but I think there is still more work to do going forward.
